ch10/sem/sem_env.go

```diff
@@ -244,7 +244,7 @@ func (ev *SemEnv) Action(element string, input tensor.Tensor) {
 // String returns the string rep of the LED env state
 func (ev *SemEnv) String() string {
 	cpar := ev.CurPara()
-	if cpar == nil {
+	if cpar == nil || len(cpar) == 0 {
```

Comment:
If it is nil, the length is guaranteed to be zero, so it is equivalent to just check if the length is zero without the nil check.

Reply:
I see, Go lets you call len() on a nil object. Good to know!